2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+would be from `reactive/relations/myapp/peers.py`, in their respective layers.
+
+Also note that those instances would be available even if their relation is not
+related to any remote application.
+
+# The SimpleRelation API
+
+The `SimpleRelation` base class will provide a couple of automatically managed
+reactive flags, as well as providing collections for introspecting the state of
+the relation, including related applications and units, relation data provided
+by those units, and for associating reactive flags with particular related
+applications or units and filtering related applications and units by those
+flags.
+
+## Managed Flags
+
+The automatically managed flags are:
+
+  * `relations.{relation_name}.joined`
+  * `relations.{relation_name}.changed`
+
+where the `{relation_name}` is replaced at runtime with the appropriate
+relation name for that instance.  (The relation name for an instance can also
+be accessed as `relation_instance.relation_name`.)
+
+The `joined` flag will set as long as at least one remote unit is related via
+the relation, and will be updated whenever the list of related units changes so
+that any handler watching the flag will be re-invoked.
+
+The `changed` flag will be set as long as at least one remote unit is
+related and has set data on the relation, and it will be updated whenever
+either a unit updates its relation data or if a related unit goes away.  The
+interface can select only units whose relation data has changed using the
+`with_flag` filter method described below and this flag.
+
+## Application and Unit Collections
+
+The interface layer can iterate over related applications and units using the
+provided collections.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        joined = 'relation.{relation_name}.joined'
+
+    @when(flags.joined)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units:
+                print('Unit {} has data: {}'.format(unit.name, unit.data))
+```
+
+Applications and units can have flags associated with them.  Flags that are
+added to an application are automatically added to all of that application's
+units, even ones that join after the fact, and if a flag is added to at least
+one of an application's units, that application will be considered to have that
+flag.
+
+You can filter the list of applications and units by flags.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        changed = 'relation.{relation_name}.changed'
+        ready = 'relation.{relation_name}.ready'
+
+    @when(flags.changed)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units.with_flag(self.flags.changed):
+                host = unit.data.host
+                print('unit {} host changed: {}'.format(unit.name, host))
+                if host:
+                    unit.add_flag(self.flags.ready)
```

Comment:
Individual relations need states, whether or not those need to be flags is perhaps up for discussion. But you definitely need to have some way of determining which relations are, for example, `ready` and which relations are not.

Not using the flag concept when writing relation interfaces but instead call it "relation state" or whatever might be just as confusing. I do see the concern about flags being used differently in charms and in relation interfaces though.

Since relation interfaces are moving towards becoming more alike layers, why not try to write them more like layers as well? Would it perhaps be possible to make the use of flags more similar in both charms and relations? They are, after all, simply on/off toggles.